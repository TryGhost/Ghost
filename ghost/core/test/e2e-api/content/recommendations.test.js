const {agentProvider, fixtureManager, matchers} = require('../../utils/e2e-framework');
const recommendationsService = require('../../../core/server/services/recommendations');
const {Recommendation} = require('@tryghost/recommendations');
const {anyObjectId, anyISODateTime} = matchers;
const assert = require('assert/strict');

describe('Recommendations Content API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getContentAPIAgent();
        await fixtureManager.init('api_keys');
        await agent.authenticate();

        // Clear placeholders
        for (const recommendation of (await recommendationsService.repository.getAll())) {
            recommendation.delete();
            await recommendationsService.repository.save(recommendation);
        }

        // Add 7 recommendations using the repository
        for (let i = 0; i < 7; i++) {
            const recommendation = Recommendation.create({
                title: `Recommendation ${i}`,
                description: `Description ${i}`,
                url: new URL(`https://recommendation${i}.com`),
                favicon: null,
                featuredImage: null,
                excerpt: null,
                oneClickSubscribe: false,
                createdAt: new Date(i * 5000) // Reliable ordering
            });

            await recommendationsService.repository.save(recommendation);
        }
    });

    it('Can paginate recommendations', async function () {
        await agent.get(`recommendations/`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': matchers.anyContentVersion,
                etag: matchers.anyEtag
            })
            .matchBodySnapshot({
                recommendations: new Array(5).fill({
                    id: anyObjectId,
                    created_at: anyISODateTime,
                    updated_at: anyISODateTime
                })
            });

        await agent.get(`recommendations/?page=2`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': matchers.anyContentVersion,
                etag: matchers.anyEtag
            })
            .matchBodySnapshot({
                recommendations: new Array(2).fill({
                    id: anyObjectId,
                    created_at: anyISODateTime,
                    updated_at: anyISODateTime
                })
            });
    });

    it('Does not allow includes', async function () {
        const {body} = await agent.get(`recommendations/?include=count.clicks,count.subscribers`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                'content-version': matchers.anyContentVersion,
                etag: matchers.anyEtag
            })
            .matchBodySnapshot({
                recommendations: new Array(5).fill({
                    id: anyObjectId,
                    created_at: anyISODateTime,
                    updated_at: anyISODateTime
                })
            });

        assert(body.recommendations[0].count === undefined);
        assert(body.recommendations[1].count === undefined);
        assert(body.recommendations[2].count === undefined);
        assert(body.recommendations[3].count === undefined);
        assert(body.recommendations[4].count === undefined);
    });
});
