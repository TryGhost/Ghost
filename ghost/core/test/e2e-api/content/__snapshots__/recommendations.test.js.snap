// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendations Content API Can paginate recommendations 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 5,
      "next": 2,
      "page": 1,
      "pages": 2,
      "prev": null,
      "total": 7,
    },
  },
  "recommendations": Array [
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 6",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 6",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation6.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 5",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 5",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation5.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 4",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 4",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation4.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 3",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 3",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation3.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 2",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 2",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation2.com/",
    },
  ],
}
`;

exports[`Recommendations Content API Can paginate recommendations 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "1545",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Recommendations Content API Can paginate recommendations 3: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 5,
      "next": null,
      "page": 2,
      "pages": 2,
      "prev": 1,
      "total": 7,
    },
  },
  "recommendations": Array [
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 1",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 1",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation1.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 0",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 0",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation0.com/",
    },
  ],
}
`;

exports[`Recommendations Content API Can paginate recommendations 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "681",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Recommendations Content API Does not allow includes 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 5,
      "next": 2,
      "page": 1,
      "pages": 2,
      "prev": null,
      "total": 7,
    },
  },
  "recommendations": Array [
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 6",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 6",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation6.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 5",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 5",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation5.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 4",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 4",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation4.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 3",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 3",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation3.com/",
    },
    Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "description": "Description 2",
      "excerpt": null,
      "favicon": null,
      "featured_image": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "one_click_subscribe": false,
      "title": "Recommendation 2",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": "https://recommendation2.com/",
    },
  ],
}
`;

exports[`Recommendations Content API Does not allow includes 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "public, max-age=0",
  "content-length": "1545",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
